### ISSUE TITLE
importless Hooks

### ISSUE THREAD

**Issue by iamneal** [tcncloud/protoc-gen-persist#75]
This is a proposed feature to make before and after hooks more flexible in where they are allowed to be written.

In the current system, a hook is defined like this:
```proto
rpc UniarySelectWithHooks(test.PartialTable) returns (test.ExampleTable) {
    option (persist.ql) = {
      query: ["SELECT * from example_table Where id=$1 AND start_time>$2"]
      arguments: ["id", "start_time"]
      before: {
        name: "UniarySelectBeforeHook"
        package: "github.com/tcncloud/protoc-gen-persist/tests/hooks"
      }
      after: {
        name: "UniarySelectAfterHook"
        package: "github.com/tcncloud/protoc-gen-persist/tests/hooks"
      }
    };
  };
```
you give the type of hook (before/after),  the name, and the package that the hook resides in.

protoc-gen-persist then generates a persist file that imports this package, in this case `"github.com/tcncloud/protoc-gen-persist/tests/hooks"` and then we can use the custom hooks in their specified locations.

The problem being the before and after hooks need to import the protobuf package that contains the messages used in the request and response of the grpc call.  If done incorrectly, it is easy to generate an import cycle.

### Solution
Remove the need to import the hooks package to use the function, by writing a function that provides the hooks by name, and give it to the builder.

In this example, the proposed code could look something like this:
```go
type AmazingImplBuilder struct {
	err           error
	rest          RestOfAmazingHandlers
	queryHandlers *persist_lib.AmazingQueryHandlers
	i             *AmazingImpl
	db            sql.DB
         hooks     AmazingImplHooks
}

func NewAmazingBuilder() *AmazingImplBuilder {
	return &AmazingImplBuilder{i: &AmazingImpl{}}
}

type AmazingImplHooks struct {
     unarySelectBeforeHook func(*Request) (*Response, error)
     unarySelectAfterHook func(*Request, *Response) error
}
func (b *AmazingImplBuilder) WithHooks(h AmazingImplHooks) { ... }
```

**Comment by iamneal** [comment]
An alternative solution to this would be to use  [this package] (https://golang.org/pkg/plugin/) 

Instead of importing types, or hooks and calling the function directly, we could 
provide a builder function like this:
```go
func NewAmazingBuilder() *AmazingImplBuilder {
	return &AmazingImplBuilder{i: &AmazingImpl{}}
}
func (b *AmazingImplBuilder) WithPlugin(path string) { 
    p, _ := plugin.Open(path)
    b.plugin = p
}
```

when calling before/after hooks, we could generate code that looks like this:  (with hook name unarySelectBeforeHook)

```go
func (s *AmazingImpl) UniarySelectWithHooks(ctx context.Context, req *test.PartialTable) (*test.ExampleTable, error) {
	var err error
	var res = &test.ExampleTable{}
	_ = err
	_ = res
        // lookup our hook by the name given on this persist option
        beforeHook, _  = s.plugin.Lookup("UnarySelectBeforeHook")

        // transform pointer to the correct function signature, and call the function with the request
	beforeRes, err := beforeHook.(func(*Request)(*Response, error)(req)
	if err != nil {
		return nil, gstatus.Errorf(codes.Unknown, "error in before hook: %v", err)
	} else if beforeRes != nil {
		return beforeRes, nil
	}
```
(old way for context)
```go
func (s *AmazingImpl) UniarySelectWithHooks(ctx context.Context, req *test.PartialTable) (*test.ExampleTable, error) {
	var err error
	var res = &test.ExampleTable{}
	_ = err
	_ = res
	beforeRes, err := hooks.UniarySelectBeforeHook(req)
	if err != nil {
		return nil, gstatus.Errorf(codes.Unknown, "error in before hook: %v", err)
	} else if beforeRes != nil {
		return beforeRes, nil
	}
```

The pros of this way over the proposed way is that we can swap out compiled code at runtime, without recompiling the program.  All that would be required to change the behavior of a hook at runtime would be to swap out the plugin with the same name, and restart the program.

One of the cons would be that before/after hooks would not be able to be supported on windows, since plugins are not available there.

Either way would require about the same amount of time to implement

Opinions @namtzigla ?

**Comment by namtzigla** [comment]
this is closed by the PR #82
